### PR TITLE
fix: Colorizer freezes on missing frame or backdrop

### LINKED ISSUE #143: Interface freezes if a dataset is missing frames
## Description

- If a user attempts to open a dataset that has metadata but no frames, the interface will stall and they will be unable to select another dataset.

image

The error emits from this line:
```
  const handleDatasetChange = useCallback(
    async (newDatasetKey: string): Promise<void> => {
      if (newDatasetKey !== datasetKey && collection) {
        const result = await collection.tryLoadDataset(newDatasetKey);
        if (result.loaded) {
          await replaceDataset(result.dataset, newDatasetKey);
        } else {
          // TODO: What happens when you try to load a bad dataset from the dropdown? Notifications?
          console.error(result.errorMessage);
        }
      }
    },  //< !!!! Uncaught (in promise) Event
    [replaceDataset, collection, datasetKey]
  );
  ```

## Expected Behavior

- Users should be warned that a dataset is missing some information and sent back to the last known good dataset.

### PR DESCRIPTION
**Pull request by ShrimpCryptid**
Problem
=======
Closes #143. The colorizer UI used to freeze playback if a frame could not be loaded, which can happen if the user doesn't have network access to the files or if the files have been deleted.

*Estimated review size: small, 10 minutes*

Solution
========
- `ColorizeCanvas` can now handle failed promises, showing a blank frame or backdrop and allowing playback to resume.
- Adds an icon to the canvas when an expected backdrop or frame image is missing.

## Type of change
* Bug fix (non-breaking change which fixes an issue)
* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------
Missing frames:
1. Open the current build and start playback. The interface should freeze at frame 25/26: 
https://dev-aics-dtp-001.int.allencell.org/nucmorph-colorizer/dist/index.html#/viewer?collection=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fassay-dev%2Fcomputational%2Fdata%2FEMT_deliverable_processing%2FTranscription_factor_cell_lines%2FVersion_1%2Fcolorizer_output%2FEOMES%2Fcollections_all.json&dataset=AICS-122_3D-MG-EMT-no-MG_fms_id_592af19bfc414ae1bf05b81967190944
3. Open the preview link for this fix and start playback. The interface should continue playing past frame 26: https://allen-cell-animated.github.io/nucmorph-colorizer/pr-preview/pr-284/#/viewer?collection=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fassay-dev%2Fcomputational%2Fdata%2FEMT_deliverable_processing%2FTranscription_factor_cell_lines%2FVersion_1%2Fcolorizer_output%2FEOMES%2Fcollections_all.json&dataset=AICS-122_3D-MG-EMT-no-MG_fms_id_592af19bfc414ae1bf05b81967190944

Missing backdrops:
1. Open the current build and start playback. The interface should freeze at frame 10: https://dev-aics-dtp-001.int.allencell.org/nucmorph-colorizer/dist/index.html#/viewer?dataset=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fdan-data%2Fcolorizer%2Fdata%2Ftest%2F3500005828_25%2Fmanifest.json&keep-range=0&bg-key=ch0ch1ch2
2. Open the preview link for this fix and start playback. The interface should continue playing past frame 10: https://allen-cell-animated.github.io/nucmorph-colorizer/pr-preview/pr-284/#/viewer?dataset=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fdan-data%2Fcolorizer%2Fdata%2Ftest%2F3500005828_25%2Fmanifest.json&keep-range=0&bg-key=ch0ch1ch2

Screenshots (optional):
-----------------------
image

image

### CODE REVIEW

File: src/colorizer/ColorizeCanvas.ts [hunk 1]
```diff
@@ -552,23 +560,48 @@ export default class ColorizeCanvas {
       backdropPromise = this.dataset?.loadBackdrop(this.selectedBackdropKey, index);
     }
     const framePromise = this.dataset?.loadFrame(index);
-    const result = await Promise.all([framePromise, backdropPromise]);
+    const result = await Promise.allSettled([framePromise, backdropPromise]);
```
Comment (ShrimpCryptid) [inline src/colorizer/ColorizeCanvas.ts:563]: `allSettled` allows promises to fail.

File: src/colorizer/Dataset.ts [hunk 2]
```diff
@@ -269,6 +269,11 @@ export default class Dataset {
       return cachedFrame;
     }
 
+    // Allow for undefined or null frame files in the manifest
+    if (this.frameFiles[index] === undefined || this.frameFiles[index] === null) {
```
Comment (toloudis) [inline src/colorizer/Dataset.ts:273]: empty string too maybe?

File: src/components/CanvasWrapper.tsx [hunk 3]
```diff
@@ -1,13 +1,31 @@
...
 
+const MissingFileIconContainer = styled(FlexColumnAlignCenter)`
+  position: absolute;
+  top: 50%;
+  left: 50%;
+  transform: translate(-50%, -50%);
+  background-color: #fff9;
+  padding: 10px 15px;
+  border-radius: 6px;
+  // TODO: Make this responsive to background color?
```
Comment (toloudis) [inline src/components/CanvasWrapper.tsx:22]: opacity...?

**Review by toloudis — APPROVED** [review]
LGTM!

**Comment by rugeli** [comment]
I couldn't load the dataset to start playback, should I use vpn? Is there a specific vpn we use? 
Error msg FYI `Error: Could not load dataset manifest 'https://dev-aics-dtp-001.int.allencell.org/dan-data/colorizer/data/test/3500005828_25/manifest.json'. ("TypeError: Failed to fetch")`

**Review by rugeli — APPROVED** [review]

**Pull request merged**